pilot/pkg/model/push_context.go

```diff
@@ -1370,17 +1370,19 @@ func (ps *PushContext) initServiceRegistry(env *Environment) {
 	// Sort the services in order of creation.
 	allServices := SortServicesByCreationTime(env.Services())
 	for _, s := range allServices {
-		svcKey := s.Key()
-		// Precache instances
+		portMap := map[string]int{}
 		for _, port := range s.Ports {
-			if _, ok := ps.ServiceIndex.instancesByPort[svcKey]; !ok {
-				ps.ServiceIndex.instancesByPort[svcKey] = make(map[int][]*ServiceInstance)
-			}
-			instances := make([]*ServiceInstance, 0)
-			instances = append(instances, env.InstancesByPort(s, port.Port)...)
-			ps.ServiceIndex.instancesByPort[svcKey][port.Port] = instances
+			portMap[port.Name] = port.Port
 		}
 
+		svcKey := s.Key()
+		if _, ok := ps.ServiceIndex.instancesByPort[svcKey]; !ok {
+			ps.ServiceIndex.instancesByPort[svcKey] = make(map[int][]*IstioEndpoint)
+		}
+		shards, ok := env.EndpointIndex.ShardsForService(string(s.Hostname), s.Attributes.Namespace)
+		if ok {
+			ps.ServiceIndex.instancesByPort[svcKey] = shards.CopyEndpoints(portMap)
```

Comment:
from here it seems we donot need port name

Reply:
See comment above, port name is the only reference to service port in endpoint.